### PR TITLE
Adding meta tag for blog post

### LINKED ISSUE #692: Meta tag for blog social media preview
## Currently when we paste the link to [https://acmcsuf.com/blog/643](https://acmcsuf.com/blog/643) in Discord or any other social media, it'll look boring like this:

image

## We want something cool like this so it's easier to share the latest blog post and more eye-catching for members

image

## This may be helpful, [Open Graph Meta Tag](https://www.opengraph.xyz/url/https%3A%2F%2Facmcsuf.com%2Fblog%2F643)

### PR DESCRIPTION
**Pull request by anhduy1202**
# Description
- I've added meta tag for blog post such that if user copy paste blog link, it should display an embedded preview

# Note:
- I've installed a package  [svelte-meta-tags](https://github.com/oekazuma/svelte-meta-tags) according to Svelte recommendation that help displaying the meta tag 

# Screenshots:

Screen Shot 2022-11-08 at 3 53 55 AM

### CODE REVIEW

**Comment by anhduy1202** [comment]
look like the DOMParser() caused the 500 error that fail website check, will fix it tmr

File: src/routes/(site)/blog/[id]/blog-body.svelte [hunk 1]
```diff
@@ -30,6 +34,11 @@
   });
 </script>
 
+<MetaTags
+  openGraph={{
+    description: `${description}`,
+  }}
+/>
```
Comment (EthanThatOneKid) [inline src/routes/(site)/blog/[id]/blog-body.svelte]: `MetaTags` usage in `src/routes/(site)/blog/[id]/blog-body.svelte` may be redundant as it is already in use on the page endpoint `src/routes/(site)/blog/[id]/+page.svelte`.

File: src/routes/(site)/blog/[id]/+page.svelte [hunk 2]
```diff
@@ -5,9 +5,23 @@
...
 
+<MetaTags
+  openGraph={{
+    title: `${data.post.title}`,
+    url: `${data.post.url}`,
+    type: 'article',
+    article: {
+      publishedTime: `${data.post.createdAt}`,
+      modifiedTime: `${data.post.lastEdited}`,
+      authors: [`${data.post.author.url}`],
```
Comment (EthanThatOneKid) [inline src/routes/(site)/blog/[id]/+page.svelte]: Unless I am missing something, the template literals are optional.
```suggestion
    title: data.post.title,
    url: data.post.url,
    type: 'article',
    article: {
      publishedTime: data.post.createdAt,
      modifiedTime: data.post.lastEdited,
      authors: [data.post.author.url],
```

File: src/routes/(site)/blog/[id]/+page.svelte [hunk 3]
```diff
@@ -5,9 +5,23 @@
   import { Temporal } from '@js-temporal/polyfill';
   import Labels from '$lib/components/blog/labels.svelte';
   import BlogBody from './blog-body.svelte';
+  import { MetaTags } from 'svelte-meta-tags';
   export let data: PageData;
```
Comment (EthanThatOneKid) [inline src/routes/(site)/blog/[id]/+page.svelte]: Nit: Leave a space between the imports and the exports.
```suggestion
  import { MetaTags } from 'svelte-meta-tags';

  export let data: PageData;
```

**Review by EthanThatOneKid — CHANGES_REQUESTED** [review]
Thanks for kicking off this initiative. Can't wait to see relevant blog content before clicking on the blog post URL. I recommend removing issue #525 from this PR since here we are specifically implementing text meta tags as to resolve #692. Thoughts on making a separate PR to resolve #525?

**Comment by anhduy1202** [comment]
# Update
- I just refactored. deleted DOMParser() and use existed query from Github instead 

# Note
I have to use string literals here since the allowed type is string | undefined form 'svelte-meta-tag' while the existing type of data.post.lastEdited is string | null
```js
modifiedTime: `${data.post.lastEdited}`,
```

# Screenshot
Screen Shot 2022-11-08 at 9 16 26 PM

File: src/routes/(site)/blog/[id]/blog-body.svelte (hunk 1 above)
Comment (anhduy1202) [reply src/routes/(site)/blog/[id]/blog-body.svelte]: done

File: src/routes/(site)/blog/[id]/+page.svelte (hunk 2 above)
Comment (anhduy1202) [reply src/routes/(site)/blog/[id]/+page.svelte]: done

File: src/routes/(site)/blog/[id]/+page.svelte (hunk 3 above)
Comment (anhduy1202) [reply src/routes/(site)/blog/[id]/+page.svelte]: done

File: src/lib/components/blog/blog-post-preview.svelte [hunk 4]
```diff
@@ -4,7 +4,6 @@
   import { readingTime } from '$lib/public/blog/utils';
   import { Temporal } from '@js-temporal/polyfill';
   import Labels from '$lib/components/blog/labels.svelte';
-
```
Comment (jaasonw) [inline src/lib/components/blog/blog-post-preview.svelte]: why?

File: src/routes/(site)/blog/[id]/blog-body.svelte [hunk 5]
```diff
@@ -4,7 +4,6 @@
   import { copy } from '$lib/public/copy/copy';
 
   export let data = '';
-
```
Comment (jaasonw) [inline src/routes/(site)/blog/[id]/blog-body.svelte]: why?

File: src/routes/(site)/blog/[id]/+page.svelte [hunk 6]
```diff
@@ -5,9 +5,29 @@
...
+  let description: string;
+  if (data.post.bodyText) {
+    description = data.post.bodyText.substring(0, 100);
+  }
 </script>
 
+<MetaTags
+  openGraph={{
+    title: data.post.title,
+    description: description,
```
Comment (jaasonw) [inline src/routes/(site)/blog/[id]/+page.svelte]: not seem to work, can you look into why?

image

File: src/lib/public/blog/types.ts [hunk 7]
```diff
@@ -9,6 +9,7 @@ export interface BlogPost {
   discussionUrl: string;
   title: string;
   html: string;
+  bodyText?: string;
```
Comment (jaasonw) [inline src/lib/public/blog/types.ts]: how come mark optional?

**Review by jaasonw — CHANGES_REQUESTED** [review]
u cant hide from me

File: src/lib/public/blog/types.ts (hunk 7 above)
Comment (anhduy1202) [reply src/lib/public/blog/types.ts]: cause it's not required, we only need that for meta tag

File: src/routes/(site)/blog/[id]/blog-body.svelte (hunk 5 above)
Comment (EthanThatOneKid) [inline src/routes/(site)/blog/[id]/blog-body.svelte]: Nit: Undo changes in `src/routes/(site)/blog/[id]/blog-body.svelte`.

File: src/lib/public/blog/types.ts (hunk 7 above)
Comment (anhduy1202) [reply src/lib/public/blog/types.ts]: done, it's required now

File: src/lib/server/blog/posts.ts [hunk 8]
```diff
@@ -77,6 +77,7 @@ function cacheBlogPosts(output: any): BlogPost[] {
       lastEditedAt: lastEdited,
       number: id,
       bodyHTML: html,
+      bodyText: bodyText,
```
Comment (EthanThatOneKid) [inline src/lib/server/blog/posts.ts]: Nit: Redundant destructor. 
```suggestion
      bodyText,
```

File: src/lib/server/blog/data.ts [hunk 9]
```diff
@@ -1,45 +1,129 @@
 export const SAMPLE_BLOG_POSTS = [
```
Comment (EthanThatOneKid) [inline src/lib/server/blog/data.ts]: Move JSON content below to a separate JSON file. Import it here with proper type checking.

File: src/lib/components/blog/blog-post-preview.svelte [hunk 10]
```diff
@@ -4,7 +4,6 @@
   import { readingTime } from '$lib/public/blog/utils';
   import { Temporal } from '@js-temporal/polyfill';
   import Labels from '$lib/components/blog/labels.svelte';
```
Comment (EthanThatOneKid) [inline src/lib/components/blog/blog-post-preview.svelte]: Nit: Remove changes in `src/lib/components/blog/blog-post-preview.svelte`.

File: src/routes/(site)/blog/[id]/+page.svelte [hunk 11]
```diff
@@ -5,9 +5,34 @@
...
+  import { MetaTags } from 'svelte-meta-tags';
+
   export let data: PageData;
+  let description: string = data.post.bodyText.substring(0, 165) ?? '';
 </script>
 
+<MetaTags
+  openGraph={{
+    title: data.post.title,
+    description: description,
```
Comment (EthanThatOneKid) [inline src/routes/(site)/blog/[id]/+page.svelte]: Nit: Move description definition to the MetaTag initialization. Related: Truncate to the last index of a space character `‘  ‘` or place there be ellipses at the end of the truncated description string.

**Review by EthanThatOneKid — CHANGES_REQUESTED** [review]
Preview looks good on Discord so far!

0136CBB7-94BB-426E-87BF-E9C67E478986

File: src/routes/(site)/blog/[id]/+page.svelte [hunk 12]
```diff
@@ -5,9 +5,34 @@
...
+      modifiedTime: data.post.lastEdited ?? '',
+      authors: [data.post.author.fullname ?? ''],
+      tags: [...data.post.labels],
+    },
+    images: [
+      {
+        url: data.post.author.picture,
+        width: 80,
+        height: 80,
+        alt: `photo of ${data.post.author.fullname}`,
```
Comment (jaasonw) [inline src/routes/(site)/blog/[id]/+page.svelte]: capitalize p

Comment (anhduy1202) [reply src/routes/(site)/blog/[id]/+page.svelte]: tf

Comment (anhduy1202) [reply src/routes/(site)/blog/[id]/+page.svelte]: done

File: src/routes/(site)/blog/[id]/blog-body.svelte (hunk 5 above)
Comment (anhduy1202) [reply src/routes/(site)/blog/[id]/blog-body.svelte]: done

File: src/lib/server/blog/posts.ts (hunk 8 above)
Comment (anhduy1202) [reply src/lib/server/blog/posts.ts]: done

File: src/lib/components/blog/blog-post-preview.svelte (hunk 10 above)
Comment (anhduy1202) [reply src/lib/components/blog/blog-post-preview.svelte]: done

File: src/routes/(site)/blog/[id]/+page.svelte (hunk 11 above)
Comment (anhduy1202) [reply src/routes/(site)/blog/[id]/+page.svelte]: done

File: src/lib/server/blog/data.ts (hunk 9 above)
Comment (anhduy1202) [reply src/lib/server/blog/data.ts]: done

**Review by EthanThatOneKid — APPROVED** [review]
> 3D807148-7492-4872-8780-E782AF9428D6

LGTM!

**Review by jaasonw — APPROVED** [review]
ur so sexy for that

**Pull request merged**